fabtests/hpcs/src/test/onesided.c

```diff
@@ -0,0 +1,501 @@
+				ret = fi_readmsg(endpoint, &ctxinfo->rma_msg, flags);
+				break;
+			case ADD:
+				ret = fi_atomicmsg(endpoint, &ctxinfo->atomic_msg, flags);
+				break;
+			default:
+				return -FI_ENOSYS;
+				break;
+			}
+		};
+	}
+
+	return ret;
+}
+
+/* RMA memory region for receive is set up in advance; nothing to do. */
+static int rx_transfer(const struct test_arguments *arguments,
+		const size_t transfer_id, const size_t transfer_count,
+		const fi_addr_t tx_address, struct fid_ep *endpoint,
+		struct op_context *op_context,
+		uint8_t *buffer, void *desc, uint64_t flags)
+{
+	return 0;
+}
+
+static int datacheck(const struct test_arguments *arguments,
+		const uint8_t *buffer, uint8_t expected, size_t len)
+{
+	int our_ret = FI_SUCCESS;
+	size_t b;
+
+	for (b = 0; b < len; b += 1) {
+		if (((uint8_t *) buffer)[b] != expected) {
+			fprintf (stderr, "datacheck failed at byte %zu (expected %u, got %u)\n",
+					b, expected, ((uint8_t *) buffer)[b]);
+			our_ret = -FI_EIO;
+			goto err_data_check;
+		}
+	}
+
+err_data_check:
+	return our_ret;
+}
+
+static int tx_datacheck(const struct test_arguments *arguments,
+		const uint8_t *buffer, size_t len)
+{
+	uint8_t expected;
+
+	switch (arguments->test_op) {
+	case READ:
+	case WRITE:
+		expected = UINT8_MAX; break;
+	case ADD:
+		expected = 1; break;
+	default:
+		return -FI_EINVAL;
+	}
+
+	return datacheck(arguments, buffer, expected, arguments->transfer_size);
+}
+
+static int rx_datacheck(const struct test_arguments *args,
+		const uint8_t *buffer, size_t len, size_t rx_peers)
+{
+	uint8_t expected;
+
+	switch (args->test_op) {
+	case READ:
+	case WRITE:
+		expected = UINT8_MAX;
+		break;
+	case ADD:
+		expected =
+			args->test_op == ADD
+				? args->dest_mode == SHARED_DEST
+					? rx_peers * args->repeat
+					: args->repeat
+				: UINT8_MAX;
+		break;
+	default:
+		return -FI_EINVAL;
+	}
+
+	len = args->dest_mode == SHARED_DEST
+		? args->transfer_size
+		: args->transfer_size * rx_peers;
+
+	return datacheck(args, buffer, expected, len);
+}
+
+struct test_api test_api(void)
+{
+	struct test_api api = {
+		.parse_arguments = &parse_arguments,
+		.free_arguments = &free_arguments,
+
+		.config = &config,
+
+		.tx_init_buffer = &tx_init_buffer,
+		.rx_init_buffer = &rx_init_buffer,
+		.tx_create_mr = &test_generic_tx_create_mr,
+		.rx_create_mr = &rx_create_mr,
+
+		.tx_window_usage = &test_generic_tx_window_usage,
+		.rx_window_usage = &test_generic_rx_window_usage,
+
+		.tx_transfer = &tx_transfer,
+		.rx_transfer = &rx_transfer,
+		.tx_cntr_completion = &test_generic_tx_cntr_completion,
+		.rx_cntr_completion = &test_generic_rx_cntr_completion,
+		.tx_cq_completion = NULL,
+		.rx_cq_completion = NULL,
+
+		.tx_datacheck = &tx_datacheck,
+		.rx_datacheck = &rx_datacheck,
+
+		.tx_fini_buffer = &test_generic_tx_fini_buffer,
+		.rx_fini_buffer = &test_generic_rx_fini_buffer,
+		.tx_destroy_mr = &test_generic_tx_destroy_mr,
+		.rx_destroy_mr = &test_generic_rx_destroy_mr
+	};
+
+	return api;
```

Comment:
Here's an example of a completely useless function because someone thought that declaring a static variable was somehow less readable/maintainable/undesirable.